pkg/provider/provider.go

```diff
@@ -251,9 +253,13 @@ func (k *kubeProvider) Diff(
 	}
 	newInputs := propMapToUnstructured(newResInputs)
 
-	// Naive replacement strategy. We will kill and recreate a resource only if the name or namespace
-	// has changed.
-	replaces, err := forceNewProperties(oldInputs.Object, newInputs.Object, oldInputs.GroupVersionKind())
+	// If the new object doesn't have a name, it was probably autonamed, so we give it the old
+	// object's name.
+	newName := newInputs.GetName()
+	replaceNameIfUnnamed(newInputs, oldInputs.GetName())
```

Comment:
I think this is in the wrong place: I'm pretty sure this needs to happen in Check.

Reply:
Ah, I see what's going on here. Initially when I mutated the inputs to have the new name, they weren't getting propagated to calls to `Diff`, `Create`, _etc_. The reason for this is that we weren't marshalling those mutations back and returning those -- we were just returning the inputs verbatim.